### ISSUE TITLE
variables of associations are not accessible in subclasses 

### ISSUE THREAD

**Issue by vahdat-ab** [umple/umple#988]
Please consider the following example,
```
class A{
  0..1 sth-- * B;
}
class B{
}
class D{
  isA A;
  0..1 sth -- * B;
}
```
It generates the following code
```
  public B getB_B(int index)
  {
    B aB = (B)bs.get(index);
    return aB;
  }
```
The variable `bs` is not accessible in the class D, so there is an error. It's a private varibale in the superclass `A`.
There several other errors in other APIs in association with the variable `bs`

**Comment by TimLethbridge** [comment]
The solution is likely to change the direct variable access to bs to instead call the getbs method.

When solving this probem search for other cases, and add tests for all cases found.

**Comment by BernardYuan** [comment]
One case I found difficult to fix is the adder function generated from the example given above, in class D, we have:
~~~Java
public boolean addB(B aB)
{
    boolean wasAdded = false;
    if (bs.contains(aB)) { return false; }
    D existingSth = aB.getSth();
    if (existingSth == null)
    {
        aB.clear_sth();
        aB.addSth(this);
    }
    else if (!this.equals(existingSth))
    {
        existingSth.removeB(aB);
        addB(aB);
    }
    else
    {
         bs.add(aB);
    }
    wasAdded = true;
    return wasAdded;
}
~~~

The issues are:
1. Class D has no access to **bs**.
2. method *getBs()* does not return the reference to bs, but a duplicate of it.
3. method *addB()* is overriden, so calling *addB()* in D invokes the method itself, rather than the *addB()* defined in class A.

I have tried to change *addB()* into *super.*, however, the templates whose names end with "_relatedSpecialization" may also be applied to non-derived classes, which makes this fix unrealistic.

**Comment by TimLethbridge** [comment]
Yikes. Tricky. These kinds of issues seem to be getting more and more complex.

In an ideal design, the addB method in B should continue to work, and not be overridden, with only calls to any necessary 'checks' made by it (e.g. for verifying multiplicity bounds) being overridden in subclasses.

I notice only trivial details that differ for the addB method in the subclass D as compared to A. Not sure why the 'clear_sth()' is needed in D, for example. Perhaps this could be put into the superclass'es method (i.e. present in all add methods) and then the override is not necessary.

Another thought would be to add a necessary protected service methods into the superclass that the subclass can use to do operations on the private variable (and not a copy of it). But that increases complexity, where as the solution above decreases complexity.

The particular case here is  a bit degenerate: The association in the subclass is identical to the association in the superclass. That would imply that actually it could be ignored (as it inherits anyway). The interesting cases are where the association with is modified to have a slightly different multiplicity (e.g. a Vehicle has any number of wheels, but in the subclass Car it is modified to have a maximum of 4 wheels -- allowing for three-wheelers!), Whatever we do, we need to test for such cases too.

**Comment by BernardYuan** [comment]
One question that is blocking me right now is that in Umple's Java Generator, when generating the code of a particular class, how could I access the information regarding the base class of this class (should it be a subclass). And if there exists an association in the subclass, how can I see whether its parent class also has associations of the same type or multiplicity?

**Comment by BernardYuan** [comment]
I have encountered another tricky case.
Given two snippets of Umple code
~~~Java
class A {
  1 -- * C;
}
class B {
  isA A;
  1 -- * C;
}
class C {
}
~~~

and 

~~~Java
class A {
  1 sth -- * C;
}
class B {
  isA A;
  1 sth -- * C;
}
class C {
}
~~~

The only difference is that in the second case, the association between A/B and C is named with one single name and in the first case the association is not named. However, they result in different behaviors in code generation. In the first case, class B has a field *cs* (a list of class C objects) overriding the *cs* inherited from A. However in the second case there is no such overriding.

My question is, are these two examples supposed to produce the same output? And if so, which behavior is expected?

**Comment by TimLethbridge** [comment]
The examples should produce the same output except that in the first case a variable on the 1 side would be aB, and on the other side it would be sth. The expected behaviour, however, is that overidding association in B would be ignored since it is identical to the inherited association. Code in subclass should only change if the multiplicities change, and multiplicities are the only thing that should change (otherwise it is a different association).

**Comment by BernardYuan** [comment]
So in the given example, neither of the two cases should have a separate *cs* in class B, and in class C, there should only be one attribute *b*, instead of both *b* and *a*?

Another question is, what is the following example supposed to behave?

~~~Java
class A {
  1 sth1 -- * C;
}
class B {
  isA A;
  1 sth2 -- * C;
}
class C {
}
~~~

**Comment by TimLethbridge** [comment]
In the last comment avoid the word 'attribute', we are talking about 'variables' here, not attributes. Attributes are at the UML modeling level. Variables are in the generated code.

So right "neither of the two cases should have a separate cs in class B" is true.

However the second part should be "and in class C, there should only be one variable anA, instead of both anA  and aB". This is different from what you wrote in a couple of ways.

As for the example the last comment, it should generate error 19 for the same reason that the code below should generate error 19. A totally separate association is created as soon as any role name differs. But there need to be differences on both sides.

```
class A {
}
class B {
  isA A;
  1 sth1 -- * C;  
  1 sth2 -- * C;
}
class C {
}
```

**Comment by TimLethbridge** [comment]
So we have agreed that the rules are:

a) If an association in a subclass has identical role names on both sides (either explicit role names or the names of classes), then 
  i. if the multiplicities are the same too, then it is the same association, and can be ignored (inherited methods will all work)
 ii. If a multiplicity is different, then it is an overiding association and specialization templates can be used.
b) else if the association has one different role name, then error 19 should be raised
c) else it is a completely different association

**Comment by BernardYuan** [comment]
I have implemented a fix for the above mentioned case a) i.

The idea is: Find the association to be ignored, and delete both the Association and the AssociationVariable from the classes at both ends of the association so that relevant code will not be generated. For Association objects this idea works fine, but the difficulty is with AssociationVariables. In UmpleClass, the *getAssociationVariables* method returns an unmodifiable collection, which makes it impossible to make any modification to the AssociationVariable method.

One possible solution is to add an "ignoredAssociationVariable" list into the UmpleClass object, and then check the list during the generation phase. Is there any other suggested solution?

**Comment by TimLethbridge** [comment]
There is API generated to remove links of any association. So even though you get an unmodifiable collection, you can still delete a link by calling the right API. That would be the best solution. Think 'Umple' not 'Java'.

Your idea of creating an 'ignoredAssociationVariable' list is problematic in several ways. Firstly it would require all generators to be aware of it, including diagram generators. Secondly, you would add an association not a list!

**Comment by BernardYuan** [comment]
I am finalizing my fix for this case:

~~~Java
class A {
  1 -- * C;
}
class B {
  isA A;
  1 -- * C;
}
class C {
}
~~~

One thing I am not sure: in this case, the association in class B should be ignored, but what about in class C, in the generated Java code of class C, should there be attribute "private B b;" and corresponding methods?  If there are those methods, how do we trigger "setB" if only class A has setC"?

**Comment by TimLethbridge** [comment]
The short answer: What should appear in class C is what would appear if class B did not exist.

In other words class C would have "private A a" as it does now. Since B is a subclass of A, any B can be put into this variable at runtime.

Class A has 'addC', but class B can do that too, as it is inherited. Class C has 'setA', and if it needs to set a B, it calls setA since any B can be used wherever and A is specified.

In conclusion. Ignoring the association in class B that is identical to the inherited association means truly ignoring it.

**Comment by BernardYuan** [comment]
There is a problem with this solution, the case mentioned before is parsed equivalently with this form:

~~~Java
class A {
}
class B {
  isA A;
}
class C {
  * -- 1 A;
  * -- 1 B;
}
~~~

Then as the solution mentioned above, the association between B and C would be ignored. However, in umple compiler, this is the exact case happening in class UmpleClass, which has association with both ConstraintTree and Precondition, where class Precondition extends ConstraintTree. And class UmpleClass would require the method "getPrecondition()", which would not be generated if the ignored association is removed.

**Comment by TimLethbridge** [comment]
Refactor the compiler first to add a role name for b case. Am travelling so can’t analyse in more detail

Timothy C Lethbridge

**Comment by TimLethbridge** [comment]
One more point. Given the above you should ensure a warning is issued when ignoring duplicate associations. 

Association matches an inherited association. Use a different role name if you need a new association for another purpose. 

Timothy C Lethbridge

**Comment by BernardYuan** [comment]
While working on the case a) ii. I found another issue with association, therefore I created another issue [issue#1238](https://github.com/umple/umple/issues/1238) and provided fix [PR#1239](https://github.com/umple/umple/pull/1239)

**Comment by BernardYuan** [comment]
Regarding the rules we have previously agreed on, the progress is reported as below:
a) If an association in a subclass has identical role names on both sides (either explicit role names or the names of classes), then
i. if the multiplicities are the same too, then it is the same association, and can be ignored (inherited methods will all work)
**Progress:** Resolved by [PR#1236](https://github.com/umple/umple/pull/1236), which is already merged.

ii. If a multiplicity is different, then it is an overiding association and specialization templates can be used.
**Progress:**
**1.** A bug was discovered while my investigation, explained in [issue#1238](https://github.com/umple/umple/issues/1238), fixed by [PR#1239](https://github.com/umple/umple/pull/1239), where some conflicts with mainline code need to be resolved.
**2.** There are some extra work needs to be done for this rule, which will be committed after the CI is fixed and other rules to be completed.

b) else if the association has one different role name, then error 19 should be raised
**Progress:**
Addressed by [PR#1237](https://github.com/umple/umple/pull/1237), which was merged then resulted in failures, because some umple source code violates this rule. Either should we revise this rule or modify the violating code.

c) else it is a completely different association
**Progress:** nothing needs to be done.

**Comment by BernardYuan** [comment]
Another bug with specialized association is found, regarding rule a) ii [issue#1245](https://github.com/umple/umple/issues/1245).

**Comment by TimLethbridge** [comment]
Some error 19 cases are being raised PR #1243 from the UmpleToRTCPP templates (after fixing the build in PR #1247 and #1248). Options for part b in the last comment are

b-i) To ignore cases where there is an overriding association where only one end has a different variable. Drawback for this is that is can hide bugs and would be generally confusing 

b-ii) To refactor the associations in UmpleToRTCpp where this problem exists. 

It was agreed to investigate the latter for now.

**Comment by BernardYuan** [comment]
Progress as of March 25:

The rules that we have previously agreed on have been fixed, other than one exception:

**a) If an association in a subclass has identical role names on both sides (either explicit role names or the names of classes), then if a multiplicity is different, then it is an overiding association and specialization templates can be used.**

Here the condition **if a multiplicity is different** is wrong according to Umple's definition of "specialization", a specialization makes sense only when the association in the subclass has a more specific multiplicity.

Now most of the specialization templates are no long called in an error-causing situation. The next step is to write some comprehensive test cases that covers all possible specialization cases and check the get methods.

**Comment by TimLethbridge** [comment]
Right. The Liskov Substitution Principle says that instances of a subclass must be able to be used wherever a parent class is used. So if Car isA GroundVehicle and GroundVehicle has * wheels, and Car has 4 wheels, this is valid as a specialization (more restrictive multiplicity). But the reverse is not. I think there should be a warning issued where the subclass actually broadens the constraint: 
"Multiplicity {0}  in the specialization association between {1} and {2} is less restrictive than multiplicity {4} in the inherited association with class {5}. This violates the Liskov Substitution Principle". 

The code generated would still work, so this should only be a warning. But it would be a useful warning  in practice and for educational purposes.

**Comment by TimLethbridge** [comment]
The following are some cases where the LSP would be violated

A multiplicity of * (i.e. 0..*) can be specialized to anything, including 1..*, 0..2, 0..1, 1

A multiplicity of a fixed number 2 (i.e. 2..2) can't be specialized

A multiplicity of 0..1 can be specialized to 1

A multiplicity of 1..* can be specialized to anything except 0..1

And we have to consider both sides, so
If the association in superclasses  is
association {* A -- * B}

Then essentially anything is valid as a specialization. A sample of specializations that are valid would be
1 -- 1
0..1 -- 1 
2 - 1
* -- 1
0..1 -- *
1 -- *

If the association in superclasses  is
association {1 C -- 0..1 D}

Then a inly valid specilization would be
1 -- 1

**Comment by BernardYuan** [comment]
What blocks me now is the definition of "specialization" and "relatedSpecialization". And when each of these templates should be instantiated.

**Comment by TimLethbridge** [comment]
To figure this out take a look at the logic that determines which templates will be emitted. For example, the relatedSpecialization templates to be emitted are determined in UmpleToJava/UmpleTLTemplates/association_set_specialization_reqCommonCode.ump
(ignore the .jet, just look at the prefix for the includefile and includefile2 to determine what will be emitted.

The above in turn is invoked by JavaSpecGenerator.ump

**Comment by BernardYuan** [comment]
Progress April 13, 2018:

I have located the bug which triggers the wrong templates and implemented my fix, however, the fix breaks some existing test cases. Will work on fixing the failure and update the thread later today or tomorrow.

**Comment by BernardYuan** [comment]
The issue that in some classes the wrong templates are trigger is fixed by this pull request:
[PR1258](https://github.com/umple/umple/pull/1258)

**Comment by BernardYuan** [comment]
As [PR1258](https://github.com/umple/umple/pull/1258) is merged. This issue can be identified as fully resolved and can be closed.
